Silk.NET.Vulkan.Extensions.ImGui

bmdub opened the pull request:
This is an attempt to integrate my port of imgui_impl_vulkan.cpp into Silk as "Silk.NET.Vulkan.Extensions.ImGui".

Some notes thus far:
- I'm a bit troubled by how many parameters are needed to construct this ImGuiController class.  Kinda ugly.  Thanks, Vulkan.
- The example app seems to have an issue where the imgui windows will disappear while you are shrinking the window.  Dunno what's going on there / haven't dug into it.
- I'll be out tomorrow but feel free to comment whenever you have time

Thanks!

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481) before we can accept your contribution.<br/><br/>:x: bmdub [sign now](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/Silk.NET?pullRequest=481) it.</sub>

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs:
```diff
@@ -0,0 +1,435 @@
+		{
+			int framebufferWidth = (int)(drawDataPtr.DisplaySize.X * drawDataPtr.FramebufferScale.X);
+			int framebufferHeight = (int)(drawDataPtr.DisplaySize.Y * drawDataPtr.FramebufferScale.Y);
+			if (framebufferWidth <= 0 || framebufferHeight <= 0)
+				return;
```
Fix branchless ifs in whole file. Generally, I prefer to have branches even for single-line ifs mainly to eliminate the potential for bugs.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Doesn't start with the standard file header.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs:
```diff
@@ -0,0 +1,435 @@
+using System;
```
Doesn't start with the standard file header (fix this in all files you've added, even the GLSL ones)

Perksey commented on src/Lab/ImGuiVulkan/ImGuiVulkanApplication.cs:
```diff
@@ -0,0 +1,1211 @@
+{
+	public class ImGuiVulkanApplication
+	{
+		public const bool EnableValidationLayers = true;
```
After a reformat this file will be fine, we're not as fussy over lab applications so long as they work.

Perksey commented on src/Lab/ImGuiVulkan/Program.cs:
```diff
@@ -0,0 +1,25 @@
+		static void Main(string[] args)
+		{
+			new ImGuiVulkanApplication().Run();
+        }
```
Formatting is super wonky here.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+// https://github.com/ocornut/imgui/blob/master/backends/imgui_impl_vulkan.cpp
+// (latest commit: https://github.com/ocornut/imgui/commit/6d3a980f38d849e9eac22180dce12f4c2e16794c)
+//
+// The names, layout, and overall convention was preserved as much as possible (almost), in order to make
+// future updates easy to apply.
+// Exceptions: Some unused methods were removed, as well as unused members of ImGui_ImplVulkan_InitInfo.
```
Would rather align this with the rest of the codebase, with C#-style names and everything else.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+// The aim of imgui_impl_vulkan.h/.cpp is to be usable in your engine without any modification.
+// IF YOU FEEL YOU NEED TO MAKE ANY CHANGE TO THIS CODE, please share them and your feedback at https://github.com/ocornut/imgui/
+
+// Important note to the reader who wish to integrate imgui_impl_vulkan.cpp/.h in their own engine/app.
+// - Common ImGui_ImplVulkan_XXX functions and structures are used to interface with imgui_impl_vulkan.cpp/.h.
+//   You will use those if you want to use this rendering backend in your engine/app.
+// - Helper ImGui_ImplVulkanH_XXX functions and structures are only used by this example (main.cpp) and by
+//   the backend itself (imgui_impl_vulkan.cpp), but should PROBABLY NOT be used by your own engine/app code.
+// Read comments in imgui_impl_vulkan.h.
+
+// CHANGELOG
+// (minor and older changes stripped away, please see git history for details)
+//  2021-03-22: Vulkan: Fix mapped memory validation error when buffer sizes are not multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize.
+//  2021-02-18: Vulkan: Change blending equation to preserve alpha in output buffer.
+//  2021-01-27: Vulkan: Added support for custom function load and IMGUI_IMPL_VULKAN_NO_PROTOTYPES by using ImGui_ImplVulkan_LoadFunctions().
+//  2020-11-11: Vulkan: Added support for specifying which subpass to reference during VkPipeline creation.
+//  2020-09-07: Vulkan: Added VkPipeline parameter to ImGui_ImplVulkan_RenderDrawData (default to one passed to ImGui_ImplVulkan_Init).
+//  2020-05-04: Vulkan: Fixed crash if initial frame has no vertices.
+//  2020-04-26: Vulkan: Fixed edge case where render callbacks wouldn't be called if the ImDrawData didn't have vertices.
+//  2019-08-01: Vulkan: Added support for specifying multisample count. Set ImGui_ImplVulkan_InitInfo::MSAASamples to one of the VkSampleCountFlagBits values to use, default is non-multisampled as before.
+//  2019-05-29: Vulkan: Added support for large mesh (64K+ vertices), enable ImGuiBackendFlags_RendererHasVtxOffset flag.
+//  2019-04-30: Vulkan: Added support for special ImDrawCallback_ResetRenderState callback to reset render state.
+//  2019-04-04: *BREAKING CHANGE*: Vulkan: Added ImageCount/MinImageCount fields in ImGui_ImplVulkan_InitInfo, required for initialization (was previously a hard #define IMGUI_VK_QUEUED_FRAMES 2). Added ImGui_ImplVulkan_SetMinImageCount().
+//  2019-04-04: Vulkan: Added VkInstance argument to ImGui_ImplVulkanH_CreateWindow() optional helper.
+//  2019-04-04: Vulkan: Avoid passing negative coordinates to vkCmdSetScissor, which debug validation layers do not like.
+//  2019-04-01: Vulkan: Support for 32-bit index buffer (#define ImDrawIdx unsigned int).
+//  2019-02-16: Vulkan: Viewport and clipping rectangles correctly using draw_data->FramebufferScale to allow retina display.
+//  2018-11-30: Misc: Setting up io.BackendRendererName so it can be displayed in the About Window.
+//  2018-08-25: Vulkan: Fixed mishandled VkSurfaceCapabilitiesKHR::maxImageCount=0 case.
+//  2018-06-22: Inverted the parameters to ImGui_ImplVulkan_RenderDrawData() to be consistent with other backends.
+//  2018-06-08: Misc: Extracted imgui_impl_vulkan.cpp/.h away from the old combined GLFW+Vulkan example.
+//  2018-06-08: Vulkan: Use draw_data->DisplayPos and draw_data->DisplaySize to setup projection matrix and clipping rectangle.
+//  2018-03-03: Vulkan: Various refactor, created a couple of ImGui_ImplVulkanH_XXX helper that the example can use and that viewport support will use.
+//  2018-03-01: Vulkan: Renamed ImGui_ImplVulkan_Init_Info to ImGui_ImplVulkan_InitInfo and fields to match more closely Vulkan terminology.
+//  2018-02-16: Misc: Obsoleted the io.RenderDrawListsFn callback, ImGui_ImplVulkan_Render() calls ImGui_ImplVulkan_RenderDrawData() itself.
+//  2018-02-06: Misc: Removed call to ImGui::Shutdown() which is not available from 1.60 WIP, user needs to call CreateContext/DestroyContext themselves.
+//  2017-05-15: Vulkan: Fix scissor offset being negative. Fix new Vulkan validation warnings. Set required depth member for buffer image copy.
+//  2016-11-13: Vulkan: Fix validation layer warnings and errors and redeclare gl_PerVertex.
+//  2016-10-18: Vulkan: Add location decorators & change to use structs as in/out in glsl, update embedded spv (produced with glslangValidator -x). Null the released resources.
+//  2016-08-27: Vulkan: Fix Vulkan example for use when a depth buffer is active.
```
Is this really necessary

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+		};
+
+		// Vulkan data
+		static ImGui_ImplVulkan_InitInfo g_VulkanInitInfo = new ImGui_ImplVulkan_InitInfo();
+		static RenderPass g_RenderPass = default;
+		static ulong g_BufferMemoryAlignment = 256;
+		static PipelineCreateFlags g_PipelineCreateFlags = 0x00;
+		static DescriptorSetLayout g_DescriptorSetLayout = default;
+		static PipelineLayout g_PipelineLayout = default;
+		static DescriptorSet g_DescriptorSet = default;
+		static Pipeline g_Pipeline = default;
+		static uint g_Subpass = 0;
+		static ShaderModule g_ShaderModuleVert;
+		static ShaderModule g_ShaderModuleFrag;
+		static bool g_FunctionsLoaded = true;
+		static ImGui_ImplVulkanH_FrameRenderBuffers[] g_frameRenderBuffers;
+		static GCHandle g_frameRenderBuffersHandle;
+		static ImGui_ImplVulkanH_Frame[] g_frames;
+		static GCHandle g_framesHandle;
+		static ImGui_ImplVulkanH_FrameSemaphores[] g_frameSemaphores;
+		static GCHandle g_frameSemaphoresHandle;
+
+		// Font data
+		static Sampler g_FontSampler = default;
+		static DeviceMemory g_FontMemory = default;
+		static Image g_FontImage = default;
+		static ImageView g_FontView = default;
+		static DeviceMemory g_UploadBufferMemory = default;
+		static Buffer g_UploadBuffer = default;
+
+		// Render buffers
+		static ImGui_ImplVulkanH_WindowRenderBuffers g_MainWindowRenderBuffers;
```
We shouldn't be capturing static state.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+		static ShaderModule g_ShaderModuleFrag;
+		static bool g_FunctionsLoaded = true;
+		static ImGui_ImplVulkanH_FrameRenderBuffers[] g_frameRenderBuffers;
+		static GCHandle g_frameRenderBuffersHandle;
```
GCHandles are very very slow and we should remove them completely.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+			if (draw_data.TotalVtxCount > 0)
+			{
+				ReadOnlySpan<Buffer> vertex_buffers = stackalloc Buffer[] { rb.VertexBuffer };
+				ulong[] vertex_offset = new ulong[] { 0 };
```
Use stackalloc. In fact just use a local.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+			}
+
+			// Setup viewport:
+			{
```
Why the bare scope?

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+
+			// Copy to Image:
+			{
+				Span<ImageMemoryBarrier> copy_barrier = stackalloc ImageMemoryBarrier[1];
```
Why can't this just be a local, no stackalloc?

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+				return;
+
+			ImGui_ImplVulkan_CreateFontSampler(vk, device, allocator);
+			fixed (Sampler* pSampler = stackalloc Sampler[1])
```
This doesn't need to be fixed.

Perksey commented on src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:
```diff
@@ -0,0 +1,1453 @@
+				binding[0].DescriptorCount = 1;
+				binding[0].StageFlags = ShaderStageFlags.ShaderStageFragmentBit;
+				binding[0].PImmutableSamplers = pSampler;
+				fixed (DescriptorSetLayoutBinding* pBinding = binding)
```
This doesn't need to be fixed.

bmdub commented:
can do!

bmdub commented:
@Perksey here you go - 

- Moved library into Lab/ImGuiVulkan app for further iteration.
- I ended up refactoring it quite a bit, replacing the straight C -> C# conversion/abomination with something more proper.  It turns out some of that C code was doing duplicate operations!
- I cleaned up probably most of what you were commenting on.  Notably, I did not touch the exceptions because I wasn't confident on exactly how they should be.
- It seems Visual Studio has changed a bunch of guids in the .sln.  I can re-push without adding to the .sln if needed.  Also, I did not see the code spacing issues, but who knows what goes on with Visual Studio.

This should be a much better handing-off point for you guys :)  I think someone who knows Vulkan better than I can still improve on this.

Perksey reviewed (approved):
Awesome, thanks. Happy to approve for further iteration down the line.
